doc/dev/continuous-integration.rst

```diff
@@ -0,0 +1,168 @@
+     jenkins_agents -> package_repos [ label = "pull build dependencies" ];
+     jenkins_agents -> pypi [ label = "pull Python packages" ];
+     jenkins_agents -> npm [ label = "pull JavaScript packages" ];
+     jenkins_agents -> apt_mirror [ label = "pull build dependencies" ];
+     jenkins_agents -> github:api [ label = "update", color = "crimson" ];
+   }
+
+Where
+
+Sepia Lab
+   `Sepia Lab`_ is a test lab used by the Ceph project. This lab offers
+   the storage and computing resources required by our CI infra.
+
+Jenkins agents
+   are a set of machines which perform the CI jobs. In this case, they
+
+   #. pull the git repo from GitHub and
+   #. rebase the pull request against the latest master
+   #. set necessary environment variables
+   #. run ``run-make-check.sh``
+
+Chacra
+   is a server offering RESTful API allowing the clients to store and
+   retrieve binary packages. It also creates the repo for uploaded
+   packages automatically. Once a certain repo is created on chacra, the
+   configured shaman server is updated as well, then we can query shaman
+   for the corresponding repo address. Chacra not only hosts ceph packages,
+   it also hosts quite
+
+Shaman
+   is a server offering RESTful API allowing the clients to query the
+   information of repos hosted by chacra nodes. Shaman is also known
+   for its `Web UI`_. But please note, shaman does not build the
+   packages, it justs offers information of the builds.
+
+.. _Sepia Lab: https://wiki.sepia.ceph.com/doku.php
+.. _Web UI: https://shaman.ceph.com
+
+build dependencies
+------------------
+
+Just like lots of other software projects, Ceph has both build-time and
+run-time dependencies. Most of time, we are inclined to use the packages
+prebuilt by the distro. But there are cases where
+
+- the necessary dependencies are either missing in the distro, or
+- their versions are too old, or
+- they are packaged without some important feature enabled.
+- we want to ensure that the version of a certain runtime dependency is
+  identical to the one we tested in our lab.
+
+No matter what the reason is, we either need to build them from source, or
+to package them as binary packages instead of using the ones shipped by the
+distro. Quite a few build-time dependencies are included as git submodules,
+but in order to avoid rebuilding these dependencies repeatedly, we pre-built
+some of them and uploaded them to our own repos. So, when performing
+``make check``, the building hosts in our CI just pull them from our internal
+repos hosting these packages instead of building them.
+
+So far, following packages are prebuilt on ubuntu focal:
+
+libboost
+    packages `boost`_. The packages' names are changed from ``libboost-*`` to
+    ``ceph-libboost-*``, and they are instead installed into ``/opt/ceph``, so
+    they don't interfere with the official ``libboost`` packages shipped by
+    distro. Its build scripts are hosted at https://github.com/tchaikov/ceph-boost.
```

Comment:
perhaps we should move this to the ceph org

Reply:
on my todo list.

Reply:
#42131